mlir/lib/Bindings/Python/IRModule.h

```diff
@@ -725,19 +733,27 @@ class PyOperation : public PyOperationBase, public BaseContextObject {
   /// parent context's live operations map, and sets the valid bit false.
   void erase();
 
-  /// Invalidate the operation.
-  void setInvalid() { valid = false; }
-
   /// Clones this operation.
   nanobind::object clone(const nanobind::object &ip);
 
+  /// Invalidate the operation.
+  void setInvalid() {
+    nanobind::ft_lock_guard lock(getContext()->liveOperationsMutex);
+    setInvalidLocked();
+  }
+  /// Like setInvalid(), but requires the liveOperations mutex to be held.
+  void setInvalidLocked() { valid = false; }
```

Comment:
dumb question: is there a way to put a runtime assert/check here that the mutex is actually held? so that there's some way for people that don't read the doc strings (...like me...) to save themselves via compiling with asserts. e.g. i'm wondering if [`nanobind::ft_mutex::lock()`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4N8nanobind8ft_mutex4lockEv) is a no-op if the mutex is already held/locked by the thread?

Reply:
That would be nice, wouldn't it.

Sadly `PyMutex_IsLocked` isn't a public CPython API; I filed https://github.com/python/cpython/issues/134009 for that.

It would probably be possible for nanobind to clone that code, much as we're doing for `PyUnstable_TryIncRef`, but let me wait until we hear back on the CPython issue before I send them a PR adding a `.is_locked()` method on the nanobind `ft_mutex`.

Reply:
> has the nice property of being present if and only if we are in GIL-disabled mode.

seems special enough to me. we can file this as a `TODO` and just do the tedious thing instead for now (rename the affected methods).

Reply:
whoops should've read everything first - guess you already did rename things. cool